Range.getClientRects should take surrogate pairs into account

rwlbuis opened the pull request:
#### 7d8be0b05d9ea1a501cb3b152033ffb86a6046f6
<pre>
Range.getClientRects should take surrogate pairs into account
<a href="https://bugs.webkit.org/show_bug.cgi?id=285023">https://bugs.webkit.org/show_bug.cgi?id=285023</a>

Reviewed by Ryosuke Niwa.

Add logic to move to surrogate pair start for Range start and past surrogate pair end for Range end in case the trailing surrogate is indexed [1].

[1] <a href="https://drafts.csswg.org/cssom-view-1/#dom-range-getclientrects">https://drafts.csswg.org/cssom-view-1/#dom-range-getclientrects</a>

* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/range-client-rects-surrogate-indexing-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/range-client-rects-surrogate-indexing.html: Added.
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::absoluteRectsForRangeInText):

Canonical link: <a href="https://commits.webkit.org/288589@main">https://commits.webkit.org/288589@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e77ee3f86e3655b03127b8d21660cc7299d1e5e4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83737 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38038 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88809 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34745 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85822 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3445 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11316 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65137 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22973 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86783 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76087 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45426 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30300 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33794 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31047 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90187 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11002 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7951 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73574 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11226 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71913 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72797 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18022 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17062 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15764 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2326 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10954 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16426 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10802 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14277 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12574 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2282,6 +2282,11 @@ static Vector<FloatRect> absoluteRectsForRangeInText(const SimpleRange& range, T
 
     auto offsetRange = characterDataOffsetRange(range, node);
+    // Move to surrogate pair start for Range start and past surrogate pair end for Range end in case the trailing surrogate is indexed.
+    if (U16_IS_TRAIL(node.data()[offsetRange.start]) && offsetRange.start > 0 && U16_IS_LEAD(node.data()[offsetRange.start - 1]))
```
Nit: offsetRange.start > 0 -> offsetRange.start

rwlbuis replied:
Fixed.

rniwa commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2282,6 +2282,11 @@ static Vector<FloatRect> absoluteRectsForRangeInText(const SimpleRange& range, T
+    // Move to surrogate pair start for Range start and past surrogate pair end for Range end in case the trailing surrogate is indexed.
+    if (U16_IS_TRAIL(node.data()[offsetRange.start]) && offsetRange.start > 0 && U16_IS_LEAD(node.data()[offsetRange.start - 1]))
+        offsetRange.start--;
+    if (U16_IS_TRAIL(node.data()[offsetRange.end]) && offsetRange.end > 0 && U16_IS_LEAD(node.data()[offsetRange.end - 1]))
```
Ditto.

rniwa replied:
How does this make sense? Should this be checking U16_IS_LEAD on offsetRange.end and U16_IS_TRAIL on offsetRange.end + 1?

rwlbuis replied:
I think that is the correct behaviour when taking into account what the spec text says:
"If the range covers a partial [typographic character unit](https://drafts.csswg.org/css-text-4/#typographic-character-unit) (e.g. half a surrogate pair or part of a grapheme cluster), the full typographic character unit must be included for the purpose of computing the bounds of the relevant [DOMRect](https://drafts.fxtf.org/geometry-1/#domrect)." So the complete surrogate pair would need to be included both at the start and end of the range.

rniwa replied:
```
U16_IS_TRAIL(node.data()[offsetRange.end]) && offsetRange.end && U16_IS_LEAD(node.data()[offsetRange.end - 1])
```

Oh right, `node.data()[offsetRange.end]` grabs the character after the end, not the one before it. But I think we still need to also check that `offsetRange.end` is less than node.date()'s length. Otherwise, we can end up reading past the end of the buffer.

rwlbuis replied:
I checked, an exception is thrown when a start/end offset is set on the DOM level. However, for some added security I added two ASSERTS checking both start and end offset values.

rniwa commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2282,6 +2282,13 @@ static Vector<FloatRect> absoluteRectsForRangeInText(const SimpleRange& range, T
 
     auto offsetRange = characterDataOffsetRange(range, node);
+    // Move to surrogate pair start for Range start and past surrogate pair end for Range end in case the trailing surrogate is indexed.
+    ASSERT(offsetRange.start < node.data().length());
```
I don't think this assertion makes sense. It's totally possible for the start to be the end of a text node:
e.g.
```js
document.body.innerHTML = 'hello';
r = document.createRange();
r.setStart(document.body.firstChild, 5);
```

rwlbuis replied:
Thanks for pointing that out.

rniwa commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2282,6 +2282,13 @@ static Vector<FloatRect> absoluteRectsForRangeInText(const SimpleRange& range, T
+    ASSERT(offsetRange.start < node.data().length());
+    if (U16_IS_TRAIL(node.data()[offsetRange.start]) && offsetRange.start && U16_IS_LEAD(node.data()[offsetRange.start - 1]))
+        offsetRange.start--;
+    ASSERT(offsetRange.end < node.data().length());
```
Ditto. Also, we need to clamp the end at the actual end of the current node since characterDataOffsetRange will return `std::numeric_limits<unsigned>::max()` when the node isn't the end container.

rwlbuis replied:
I think we can avoid doing anything at the very start and end of ranges as we are testing for the case where we are indexing into the surrogate pair, so also no need to clamp.
I added a subtest where the range consists of a different start and end text container to test the case where std::numeric_limits<unsigned>::max() is returned.

rniwa reviewed (changes requested):
I don't think we can land this as is.

rwlbuis committed:
Range.getClientRects should take surrogate pairs into account
https://bugs.webkit.org/show_bug.cgi?id=285023

Reviewed by Ryosuke Niwa.

Add logic to move to surrogate pair start for Range start and past surrogate pair end for Range end in case the trailing surrogate is indexed [1].

[1] https://drafts.csswg.org/cssom-view-1/#dom-range-getclientrects

* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/range-client-rects-surrogate-indexing-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/range-client-rects-surrogate-indexing.html: Added.
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::absoluteRectsForRangeInText):

Canonical link: https://commits.webkit.org/288589@main
